contracts/oracle/src/contract/feed.rs

```diff
@@ -127,20 +149,442 @@ pub fn try_query_alarms<OracleBase>(
+            self.stack.clear();
+            self.err = false;
+        } else {
+            let parent_idx = self
+                .stack
+                .iter()
+                .position(|parent| parent.base_ticker() == Q::TICKER);
+
+            if let Some(n) = parent_idx {
+                self.stack.truncate(n + 1);
+                self.err = false;
+            } else {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl<OracleBase, Querier> AnyVisitorPair for &mut LegCmd<OracleBase, Querier>
+where
+    OracleBase: Currency + DeserializeOwned,
+    Querier: PriceQuerier,
+{
+    type Output = Option<BasePrice<SwapGroup, OracleBase>>;
+    type Error = ContractError;
+
+    fn on<B, Q>(self) -> Result<Self::Output, Self::Error>
+    where
+        B: Currency + DeserializeOwned,
+        Q: Currency + DeserializeOwned,
+    {
+        // recovery mode
+        if self.err && !self.recover::<B, Q>() {
+            return Ok(None);
+        }
+
+        let price: Option<BasePrice<SwapGroup, OracleBase>> = loop {
+            match self
+                .stack
+                .last()
+                .map(TryInto::<Price<Q, OracleBase>>::try_into)
+            {
+                None => {
+                    debug_assert_eq!(Q::TICKER, OracleBase::TICKER);
+
+                    break self.price_querier.price::<B, OracleBase>()?;
+                }
+                Some(Ok(price_parent)) => {
+                    break self
+                        .price_querier
+                        .price::<B, Q>()?
+                        .map(|price| price * price_parent)
+                }
+                _ => {
+                    self.stack.truncate(self.stack.len() - 1);
+                }
+            }
+        }
+        .map(|price| {
+            let bprice: BasePrice<SwapGroup, OracleBase> = price.into();
+            self.stack.push(bprice.clone());
+            bprice
+        });
+
+        if price.is_none() {
+            self.err = true;
+        }
+
+        Ok(price)
+    }
+}
+
+// TODO: rename to something meaningfull
+struct ConfiguredFeeds<'a> {
+    feeds: PriceFeeds<'static>,
+    at: Timestamp,
+    total_feeders: usize,
+    storage: &'a dyn Storage,
+}
+
+trait PriceQuerier {
+    fn price<B, Q>(&self) -> Result<Option<Price<B, Q>>, ContractError>
+    where
+        B: Currency + DeserializeOwned,
+        Q: Currency + DeserializeOwned;
+}
+
+impl<'a> PriceQuerier for ConfiguredFeeds<'a> {
+    fn price<B, Q>(&self) -> Result<Option<Price<B, Q>>, ContractError>
+    where
+        B: Currency + DeserializeOwned,
+        Q: Currency + DeserializeOwned,
+    {
+        let price = self
+            .feeds
+            .price_of_feed(self.storage, self.at, self.total_feeders)
+            .map(Some)
+            .or_else(|err| match err {
+                PriceFeedsError::NoPrice() => Ok(None),
+                _ => Err(err),
+            })?;
+        Ok(price)
+    }
+}
```

Comment:
No tests for this code. Suggest extracting the logic that converts `Result<Price<>, E> -> Result<Option<Price<>>, E>` into a function to allow testing it.